File: Python/I/11-ZOO/2/dragdroplineedit.py
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

class DragDropLineEdit(QLineEdit):
    def dragEnterEvent(self, event):
        if self.text() == '':
            event.accept()
        else:
            event.ignore()

    def dropEvent(self, event):
        src = event.source()
        self.setText(src.text())
        src.setVisible(False)

File: C++/I/11-ZOO/2/dragdroplineedit.cpp
#include "dragdroplineedit.h"
#include "dragdroplabel.h"

DragDropLineEdit::DragDropLineEdit(QWidget *parent):
    QLineEdit(parent){}

void DragDropLineEdit::dragEnterEvent(QDragEnterEvent *event)
{
    if (text() == "")
        event->accept();
    else
        event->ignore();
}

void DragDropLineEdit::dropEvent(QDropEvent *event)
{
    DragDropLabel *src = (DragDropLabel*)event->source();
    setText(src->text());
    src->setVisible(false);
}
